### PR TITLE
Add check for non-zero sample of empty table in cudf::sample()

### LINKED ISSUE #23831: cudf::sample reads out of bounds on an empty input table when sampling with replacement
`cudf::sample` on an empty input table with `n > 0` and `sample_with_replacement::TRUE` reads entirely out of bounds instead of failing cleanly.

`cpp/src/copying/sample.cu` builds the random gather map with

```cpp
thrust::uniform_int_distribution<size_type> dist{0, num_rows - 1};
```

For a 0-row table that is `{0, -1}`. In the pinned Thrust implementation the distribution collapses to a constant 0 (the real-interval path becomes `[0, 0]`), so every element of the length-n transform iterator yields row 0. `detail::gather` has no empty-source guard, so the result is an n-row table whose contents are read from out-of-bounds offsets of zero-size allocations: silent garbage values or a device-side illegal access depending on allocator state.

The only existing guard is in the Python layer (`cudf.core.indexed_frame`, "Cannot take a sample larger than 0 when axis is empty."), and pylibcudf does not wrap `cudf::sample` at all, so plain libcudf C++ callers hit this with nothing between them and the UB.

Suggested fix: reject `num_rows == 0 && n > 0` when sampling with replacement, mirroring the error the Python layer already raises.

### PR DESCRIPTION
**Pull request by davidwendt**
## Description
Throws an invalid-argument error for the case where a non-zero number of samples is requested on an empty input table for `cudf::sample()`. Also adds a gtest for this case.

Closes #23831

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/src/copying/sample.cu [hunk 1]
```diff
@@ -35,8 +35,13 @@ std::unique_ptr<table> sample(table_view const& input,
...
   auto const num_rows = input.num_rows();
 
-  if ((n > num_rows) and (replacement == sample_with_replacement::FALSE)) {
-    CUDF_FAIL("If n > number of rows, then multiple sampling of the same row should be allowed");
+  if (n > num_rows) {
+    CUDF_EXPECTS(replacement == sample_with_replacement::TRUE,
+                 "If n > number of rows, then multiple sampling of the same row should be allowed",
+                 std::invalid_argument);
+    CUDF_EXPECTS(num_rows > 0,
+                 "If n > number of rows, then the input table should not be empty",
```
Comment (bdice) [inline cpp/src/copying/sample.cu]: This message is a little tricky to parse. NumPy has some wording that I like here: "input cannot be empty unless no samples are taken." Or perhaps this wording. Your call.
```suggestion
                 "A nonzero number of samples was requested but the input table has zero rows",
```

**Review by bdice — APPROVED** [review]

File: cpp/src/copying/sample.cu [hunk 2]
```diff
@@ -32,11 +32,16 @@ std::unique_ptr<table> sample(table_view const& input,
...
 {
-  CUDF_EXPECTS(n >= 0, "expected number of samples should be non-negative");
+  CUDF_EXPECTS(n >= 0, "expected number of samples should be non-negative", std::invalid_argument);
   auto const num_rows = input.num_rows();
 
-  if ((n > num_rows) and (replacement == sample_with_replacement::FALSE)) {
-    CUDF_FAIL("If n > number of rows, then multiple sampling of the same row should be allowed");
+  if (n > num_rows) {
+    CUDF_EXPECTS(replacement == sample_with_replacement::TRUE,
+                 "If n > number of rows, then multiple sampling of the same row should be allowed",
```
Comment (igorpeshansky) [inline cpp/src/copying/sample.cu:43]: [Optional] Prior phrasing, but would `should`→`must` read better? Also in line 35…

File: cpp/src/copying/sample.cu [hunk 3]
```diff
@@ -32,11 +32,16 @@ std::unique_ptr<table> sample(table_view const& input,
...
+                 std::invalid_argument);
+    CUDF_EXPECTS(num_rows > 0,
+                 "A nonzero number of samples requested but the input table has zero rows",
+                 std::invalid_argument);
```
Comment (igorpeshansky) [inline cpp/src/copying/sample.cu]: Calling `sample` on an empty input with `n > 0` and `replacement == sample_with_replacement::FALSE` would fail with "multiple sampling should be allowed", which is misleading because enabling multiple sampling would then trigger the "table has zero rows" failure. Swapping the two checks would always cause the right failure for an empty input.

It might also make sense to add a separate test for this case.

File: cpp/src/copying/sample.cu [hunk 4]
```diff
@@ -32,11 +32,16 @@ std::unique_ptr<table> sample(table_view const& input,
...
   auto const num_rows = input.num_rows();
 
-  if ((n > num_rows) and (replacement == sample_with_replacement::FALSE)) {
-    CUDF_FAIL("If n > number of rows, then multiple sampling of the same row should be allowed");
+  if (n > num_rows) {
+    CUDF_EXPECTS(replacement == sample_with_replacement::TRUE,
+                 "If n > number of rows, then multiple sampling of the same row should be allowed",
+                 std::invalid_argument);
+    CUDF_EXPECTS(num_rows > 0,
+                 "A nonzero number of samples requested but the input table has zero rows",
```
Comment (igorpeshansky) [inline cpp/src/copying/sample.cu]: Nit: you're missing the "was" from @bdice's suggestion ("A nonzero number of samples **was** requested…").

[Really optional] I know @bdice proposed it, but "input table is empty" reads (to me) a bit cleaner than "input table has zero rows"…

File: cpp/include/cudf/copying.hpp [hunk 5]
```diff
@@ -868,8 +868,9 @@ enum class sample_with_replacement : bool {
  * output:       {col1: {3, 1, 1}, col2: {8, 6, 6}}
  * @endcode
  *
- * @throws cudf::logic_error if `n` > `input.num_rows()` and `replacement` == FALSE.
- * @throws cudf::logic_error if `n` < 0.
+ * @throws std::invalid_argument if `n` < 0.
+ * @throws std::invalid_argument if `n` > `input.num_rows()` and `replacement` == FALSE.
+ * @throws std::invalid_argument if `n` > `input.num_rows()` and `input.num_rows() == 0`.
```
Comment (igorpeshansky) [inline cpp/include/cudf/copying.hpp]: Since `input.num_rows() == 0`, the first condition is just `n > 0`. The way it's phrased now makes the reader do the inference themselves, and simplifying it would match the error message.

Also, like in https://github.com/NVIDIA/cudf/pull/23841#discussion_r3868509132, I'd swap the order of these `@throws` clauses to put the empty check first.

File: cpp/tests/copying/sample_tests.cpp [hunk 6]
```diff
@@ -26,7 +26,17 @@ TEST_F(SampleTest, FailCaseRowMultipleSampling)
...
 
   EXPECT_THROW(cudf::sample(input, n_samples, cudf::sample_with_replacement::FALSE, 0),
-               cudf::logic_error);
+               std::invalid_argument);
+}
+
+TEST_F(SampleTest, FailSamplingEmptyTable)
+{
+  cudf::test::fixed_width_column_wrapper<int32_t> col1{};
+  cudf::size_type const n_samples = 10;
```
Comment (igorpeshansky) [inline cpp/tests/copying/sample_tests.cpp:35]: Should we also test that `n == 0` with an empty table returns an empty result?

File: cpp/include/cudf/copying.hpp [hunk 7]
```diff
@@ -868,8 +868,9 @@ enum class sample_with_replacement : bool {
  * output:       {col1: {3, 1, 1}, col2: {8, 6, 6}}
  * @endcode
  *
- * @throws cudf::logic_error if `n` > `input.num_rows()` and `replacement` == FALSE.
- * @throws cudf::logic_error if `n` < 0.
+ * @throws std::invalid_argument if `n` < 0.
+ * @throws std::invalid_argument if `n` > 0` and `input.num_rows() == 0`.
```
[the remark this answers is not in the document]
Comment (igorpeshansky) [reply cpp/include/cudf/copying.hpp]: Good catch, though I would use the style consistent with the previous line: 

```suggestion
 * @throws std::invalid_argument if `n` > 0 and `input.num_rows() == 0`.
```

File: cpp/include/cudf/copying.hpp [hunk 8]
```diff
@@ -868,8 +868,9 @@ enum class sample_with_replacement : bool {
  * output:       {col1: {3, 1, 1}, col2: {8, 6, 6}}
  * @endcode
  *
- * @throws cudf::logic_error if `n` > `input.num_rows()` and `replacement` == FALSE.
- * @throws cudf::logic_error if `n` < 0.
+ * @throws std::invalid_argument if `n < 0`.
+ * @throws std::invalid_argument if `n > 0` and `input.num_rows() == 0`.
+ * @throws std::invalid_argument if `n > input.num_rows()` and `replacement == FALSE`.
```
Comment (igorpeshansky) [inline cpp/include/cudf/copying.hpp:873]: [Really optional] Most of the other doxygen comments leave the operators out of the backticks (e.g., lines 575 and 771), so we might want to match that style:

```suggestion
 * @throws std::invalid_argument if `n` < 0.
 * @throws std::invalid_argument if `n` > 0 and `input.num_rows()` == 0.
 * @throws std::invalid_argument if `n` > `input.num_rows()` and `replacement` == FALSE.
```

File: cpp/tests/copying/sample_tests.cpp [hunk 9]
```diff
@@ -26,7 +26,19 @@ TEST_F(SampleTest, FailCaseRowMultipleSampling)
...
+  cudf::table_view input({col1});
+
+  EXPECT_THROW(cudf::sample(input, n_samples, cudf::sample_with_replacement::TRUE, 0),
+               std::invalid_argument);
+  EXPECT_THROW(cudf::sample(input, n_samples, cudf::sample_with_replacement::FALSE, 0),
+               std::invalid_argument);
```
Comment (igorpeshansky) [inline cpp/tests/copying/sample_tests.cpp:41]: [Really optional] These two assertions pass regardless of the order of the checks in `sample()`, because we don't actually look at the message. We could potentially add a message content assertion using GMock's `EXPECT_THAT`, e.g.:

```suggestion
  EXPECT_THAT([&]() { cudf::sample(input, n_samples, cudf::sample_with_replacement::TRUE, 0); },
              ::testing::ThrowsMessage<std::invalid_argument>(::testing::HasSubstr("zero rows")));
  EXPECT_THAT([&]() { cudf::sample(input, n_samples, cudf::sample_with_replacement::FALSE, 0); },
              ::testing::ThrowsMessage<std::invalid_argument>(::testing::HasSubstr("zero rows")));
```

However, this would be the first use of the exception matcher in libcudf, so up to you…

**Review by igorpeshansky — APPROVED** [review]
LGTM :shipit:

File: cpp/tests/copying/sample_tests.cpp (hunk 9 above)
Comment (davidwendt) [reply cpp/tests/copying/sample_tests.cpp:41]: No, we won't be doing this again.
